# Importation des données des arbres de la ville (Data Toulouse métropole) 

In [53]:
DIR="C:/Users/User/Documents/MesProjets/projet_python/AnalyseDataToulouse"

In [1]:
import sys
import subprocess

subprocess.check_call([sys.executable, "-m", "pip", "install", "pandas", "polars"])



0

In [44]:
import requests
import pandas as pd
import pyarrow
import polars as pl
from io import StringIO #traite les chaines de caractere comme des ficheirs

### Exemple de endpoint:
- Pour le genre Quercus (sans limite de sortie) :

  /catalog/datasets/arbres-urbains/exports/json?where=genre%20%3D%20%22Quercus%22&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326

In [5]:
server = "https://data.toulouse-metropole.fr/api/explore/v2.1"
endpoint = "/catalog/datasets/arbres-urbains/exports/json?where=genre%20%3D%20%22Quercus%22&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326"

## Combien de genres se trouvent dans le dataset ?


On change le endpoint:

In [101]:
endpoint = "/catalog/datasets/arbres-urbains/exports/json?select=genre&group_by=genre&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326"

In [102]:
# Ne pas toucher
api_url = server + endpoint
print(api_url)
response = requests.get(api_url)


https://data.toulouse-metropole.fr/api/explore/v2.1/catalog/datasets/arbres-urbains/exports/json?select=genre&group_by=genre&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326


In [103]:
# Taille de la liste renvoyée
print(len(response.json()))

165


In [104]:
# Affichage
type(response.json())
response.json()

[{'genre': None},
 {'genre': '0_Genre.inconnu'},
 {'genre': 'Abies'},
 {'genre': 'Acacia'},
 {'genre': 'Acer'},
 {'genre': 'Aesculus'},
 {'genre': 'Ailanthus'},
 {'genre': 'Albizia'},
 {'genre': 'Alnus'},
 {'genre': 'Amelanchier'},
 {'genre': 'Araucaria'},
 {'genre': 'Arbutus'},
 {'genre': 'Betula'},
 {'genre': 'Bignonia'},
 {'genre': 'Brahea'},
 {'genre': 'Broussonetia'},
 {'genre': 'Buddleja'},
 {'genre': 'Butia'},
 {'genre': 'Buxus'},
 {'genre': 'Callistemon'},
 {'genre': 'Calocedrus'},
 {'genre': 'Camellia'},
 {'genre': 'Carpinus'},
 {'genre': 'Carya'},
 {'genre': 'Castanea'},
 {'genre': 'Catalpa'},
 {'genre': 'Ceanothus'},
 {'genre': 'Cedrus'},
 {'genre': 'Celtis'},
 {'genre': 'Cephalotaxus'},
 {'genre': 'Ceratonia'},
 {'genre': 'Cercidyphyllum'},
 {'genre': 'Cercis'},
 {'genre': 'Chamaecyparis'},
 {'genre': 'Chamaerops'},
 {'genre': 'Chandelle'},
 {'genre': 'Chionanthus'},
 {'genre': 'Cinnamum'},
 {'genre': 'Cladastris'},
 {'genre': 'Clerodendron'},
 {'genre': 'Clerodendrum'},
 {

## Gestion des genres un par un.

### Correction  genre 0_Genre.inconnu

In [78]:
endpoint = "/catalog/datasets/arbres-urbains/exports/json?where=genre%3D%220_Genre.inconnu%22&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326"

In [79]:
# Ne pas toucher
api_url = server + endpoint
print(api_url)
response = requests.get(api_url)


https://data.toulouse-metropole.fr/api/explore/v2.1/catalog/datasets/arbres-urbains/exports/json?where=genre%3D%220_Genre.inconnu%22&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326


In [83]:
# Affichage
type(response.json())
response.json()[0:3]

[{'id': 48690,
  'longitude': 1.379471124207392,
  'latitude': 43.577283218696145,
  'famille': '0_Famille_inconnue',
  'genre': '0_Genre.inconnu',
  'espece': 'sp.',
  'cultivar_variete': None,
  'stade_dev': None,
  'forme': None,
  'hauteur_approx_forestry': None,
  'remarquable': None,
  'date_plantation': None,
  'territoire': 'OUEST',
  'quartier': 18,
  'commune': 'Toulouse',
  'code_insee': '3120',
  'geo_point_2d': {'lon': 1.379471124207392, 'lat': 43.577283218696145}},
 {'id': 42092,
  'longitude': 1.419892446099316,
  'latitude': 43.64921987892591,
  'famille': '0_Famille_inconnue',
  'genre': '0_Genre.inconnu',
  'espece': None,
  'cultivar_variete': None,
  'stade_dev': '1 - Jeune',
  'forme': '1 - Naturelle',
  'hauteur_approx_forestry': None,
  'remarquable': None,
  'date_plantation': None,
  'territoire': 'NORD',
  'quartier': 7,
  'commune': 'Toulouse',
  'code_insee': '3122',
  'geo_point_2d': {'lon': 1.419892446099316, 'lat': 43.64921987892591}},
 {'id': 42215,
  'l

In [84]:
# Taille de la liste renvoyée
print(len(response.json()))

3046


Il y a 3046 individus dont le genre n'est pas connu et renseigné 0_Genre.inconnu. Parmi eux, vérification que l'espèce ne soit également pas connue.

In [91]:
endpoint = "/catalog/datasets/arbres-urbains/exports/json?select=espece&where=genre%3D%220_Genre.inconnu%22&group_by=espece&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326"

In [92]:
# Ne pas toucher
api_url = server + endpoint
print(api_url)
response = requests.get(api_url)


https://data.toulouse-metropole.fr/api/explore/v2.1/catalog/datasets/arbres-urbains/exports/json?select=espece&where=genre%3D%220_Genre.inconnu%22&group_by=espece&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326


In [93]:
# Affichage
type(response.json())
response.json()[0:3]

[{'espece': None}, {'espece': '0_sp.'}, {'espece': 'Everest'}]

Le pommier "Evereste" fait partie des espèces dont le genre n'a pas été renseigné?

### Genre "None"

In [12]:
endpoint = "/catalog/datasets/arbres-urbains/exports/json?where=genre%20is%20null&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326"

In [13]:
# Ne pas toucher
api_url = server + endpoint
print(api_url)
response = requests.get(api_url)
data = response.json()

https://data.toulouse-metropole.fr/api/explore/v2.1/catalog/datasets/arbres-urbains/exports/json?where=genre%20is%20null&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326


In [16]:
# Taille de la liste renvoyée
print(len(data))

33


In [17]:
# Affichage
type(data)
data

[{'id': 1106,
  'longitude': 1.3896296796262932,
  'latitude': 43.59262907739858,
  'famille': None,
  'genre': None,
  'espece': None,
  'cultivar_variete': None,
  'stade_dev': None,
  'forme': None,
  'hauteur_approx_forestry': None,
  'remarquable': None,
  'date_plantation': None,
  'territoire': 'OUEST',
  'quartier': 18,
  'commune': 'Toulouse',
  'code_insee': '3115',
  'geo_point_2d': {'lon': 1.3896296796262932, 'lat': 43.59262907739858}},
 {'id': 22478,
  'longitude': 1.4323303003667787,
  'latitude': 43.592520284766096,
  'famille': None,
  'genre': None,
  'espece': None,
  'cultivar_variete': None,
  'stade_dev': None,
  'forme': None,
  'hauteur_approx_forestry': None,
  'remarquable': None,
  'date_plantation': None,
  'territoire': 'SUD',
  'quartier': 6,
  'commune': 'Toulouse',
  'code_insee': '3115',
  'geo_point_2d': {'lon': 1.4323303003667787, 'lat': 43.592520284766096}},
 {'id': 22591,
  'longitude': 1.4290152738980775,
  'latitude': 43.595337767164075,
  'famille

In [18]:
df = pl.DataFrame(data)
print(df.select(["quartier","espece"]))


shape: (33, 2)
┌──────────┬────────┐
│ quartier ┆ espece │
│ ---      ┆ ---    │
│ i64      ┆ null   │
╞══════════╪════════╡
│ 18       ┆ null   │
│ 6        ┆ null   │
│ 6        ┆ null   │
│ 6        ┆ null   │
│ 6        ┆ null   │
│ …        ┆ …      │
│ 17       ┆ null   │
│ 6        ┆ null   │
│ 18       ┆ null   │
│ 9        ┆ null   │
│ 18       ┆ null   │
└──────────┴────────┘


In [19]:
print(df.select(["quartier","espece"]).unique())


shape: (6, 2)
┌──────────┬────────┐
│ quartier ┆ espece │
│ ---      ┆ ---    │
│ i64      ┆ null   │
╞══════════╪════════╡
│ 3        ┆ null   │
│ 5        ┆ null   │
│ 17       ┆ null   │
│ 6        ┆ null   │
│ 18       ┆ null   │
│ 9        ┆ null   │
└──────────┴────────┘


Donc au vu du peu d'espèces mal renseignées (environ 3000 sur 120000), soit moins de 3%. On peut ne pas les utiliser.

# Création du dataframe sans les Genres mal renseignés

In [20]:
endpoint = "/catalog/datasets/arbres-urbains/exports/json?where=genre%20is%20not%20null%20AND%20NOT%20genre%3D%220_Genre.inconnu%22&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326"

In [21]:
# Ne pas toucher
api_url = server + endpoint
print(api_url)
response = requests.get(api_url)


https://data.toulouse-metropole.fr/api/explore/v2.1/catalog/datasets/arbres-urbains/exports/json?where=genre%20is%20not%20null%20AND%20NOT%20genre%3D%220_Genre.inconnu%22&limit=-1&timezone=UTC&use_labels=false&compressed=false&epsg=4326


In [22]:
# Affichage
type(response.json())
response.json()
data = response.json()

In [23]:
print(len(data))

121135


In [24]:
print(data[1])

{'id': 48665, 'longitude': 1.3790883668757472, 'latitude': 43.577605029900525, 'famille': 'Fagacees', 'genre': 'Quercus', 'espece': 'rubra', 'cultivar_variete': None, 'stade_dev': '2 - Jeune adulte', 'forme': '2 - Libre', 'hauteur_approx_forestry': None, 'remarquable': None, 'date_plantation': None, 'territoire': 'OUEST', 'quartier': 18, 'commune': 'Toulouse', 'code_insee': '3120', 'geo_point_2d': {'lon': 1.3790883668757472, 'lat': 43.577605029900525}}


In [32]:
#Création du Panda dataframe
df_pd = pd.DataFrame(data)

In [33]:
print(df_pd.tail(5))

           id  longitude   latitude      famille   genre      espece  \
121130  51650   1.394111  43.583041    Malvacees   Tilia  x.europaea   
121131  51654   1.393278  43.582888    Malvacees   Tilia  x.europaea   
121132  51656   1.393232  43.582835    Malvacees   Tilia  x.europaea   
121133  51689   1.398670  43.583979     Ulmacees   Ulmus      laevis   
121134  51690   1.398718  43.583830  Cannabacees  Celtis   australis   

       cultivar_variete         stade_dev                               forme  \
121130              NaN  2 - Jeune adulte                           2 - Libre   
121131              NaN  2 - Jeune adulte                           2 - Libre   
121132              NaN  2 - Jeune adulte  4 - Architecturée en têtes de chat   
121133              NaN  2 - Jeune adulte                      3 - Semi libre   
121134              NaN  2 - Jeune adulte                           2 - Libre   

       hauteur_approx_forestry remarquable date_plantation territoire  \
121130 

In [41]:
# Changement de type des colonnes problématiques
for col in ["forme","date_plantation"]:
    df_pd[col] = df_pd[col].astype(str)  # None devient 'None'

In [42]:
# Vérif des types 
df_pd.info()

<class 'pandas.DataFrame'>
RangeIndex: 121135 entries, 0 to 121134
Data columns (total 17 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   id                       121135 non-null  int64  
 1   longitude                121135 non-null  float64
 2   latitude                 121135 non-null  float64
 3   famille                  120822 non-null  str    
 4   genre                    121135 non-null  str    
 5   espece                   121082 non-null  str    
 6   cultivar_variete         5931 non-null    str    
 7   stade_dev                119564 non-null  str    
 8   forme                    118706 non-null  str    
 9   hauteur_approx_forestry  54475 non-null   str    
 10  remarquable              37 non-null      str    
 11  date_plantation          5107 non-null    str    
 12  territoire               120306 non-null  str    
 13  quartier                 114785 non-null  float64
 14  commune        

In [55]:
# Sauvegarde en parquet
path = DIR + "/data/arbres_panda.csv"
df_pd.to_csv(path, index=False, na_rep="NA")


In [45]:
df_pl = pl.from_pandas(df_pd)

ImportError: pyarrow is required for converting a pandas dataframe to Polars, unless each of its columns is a simple numpy-backed one (e.g. 'int64', 'bool', 'float32' - not 'Int64')

In [40]:
print(df.select(["date_plantation","espece"]).unique())

shape: (1, 2)
┌─────────────────┬────────┐
│ date_plantation ┆ espece │
│ ---             ┆ ---    │
│ null            ┆ null   │
╞═════════════════╪════════╡
│ null            ┆ null   │
└─────────────────┴────────┘
